config/crd/bases/mutations.gatekeeper.sh_assign.yaml

```diff
@@ -509,11 +513,13 @@ spec:
                               items:
                                 type: string
                               type: array
+                              x-kubernetes-list-type: atomic
                           required:
                           - key
                           - operator
                           type: object
                         type: array
+                        x-kubernetes-list-type: atomic
```

Comment:
Is adding this backwards compatible?

Reply:
@maxsmythe I think so, [this page](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) says `atomic can apply to any list`. It is coming through bumping k8s api - https://github.com/kubernetes/apimachinery/blob/v0.30.3/pkg/apis/meta/v1/types.go#L1250

Reply:
Backwards compatibility means that applying this tag will not change merge behavior.

In other words, we should be sure that `kubectl apply` prior to this tag being added and post this tag being added are unchanged.

Because this potentially affects backwards compatibility, we should be certain that this is the case. We should not rely on speculation. In other words, we should do more research than just inferring from a vague statement in the docs.

Reply:
I found `By default, Server-Side Apply treats custom resources as unstructured data. All keys are treated the same as struct fields, and all lists are considered atomic.`, [here](https://kubernetes.io/docs/reference/using-api/server-side-apply/#custom-resources-and-server-side-apply). So before this change, the list was being treated as `atomic` for `patchStratagy=merge`. wdyt? 

How do I test for backward compatibility for this change?

I tested with deploying CRDs before this change and creating mutators, then I upgraded the CRDs with this change and applied mutators But I think this may be wrong way to test for backward compat, I am not sure.

Reply:
I also see this `If listType is missing, the API server interprets a patchStrategy=merge marker as a listType=map and the corresponding patchMergeKey marker as a listMapKey.` on the same page, so now I am cofused.

Reply:
I asked in slack - https://kubernetes.slack.com/archives/C0123CNN8F3/p1722991068800879, waiting for response.

Reply:
I got the response - `atomic is the default` - https://kubernetes.slack.com/archives/C0123CNN8F3/p1722991724774759?thread_ts=1722991068.800879&cid=C0123CNN8F3.

@maxsmythe let me know if this resolves your concerns.

Reply:
I am pasting the response for reference here about this change - 

```
atomic is the default
that was making it explicit to avoid linting errors about implicit list type
```
```
the containing labelselector struct is already annotated as +structType=atomic, which generators should turn into an atomic annotation on the parent selector field, so no list type annotations on the child fields should be relevant
```

Reply:
Adding more context:

This `x-kubernetes-list-type: atomic` is getting added for list fields under `metav1.labelSelector` and `metav1.namespaceSelector` for `mutators`. This change is coming from [here](https://github.com/kubernetes/apimachinery/blob/v0.30.3/pkg/apis/meta/v1/types.go#L1250), where marker `listType=atomic` is getting added to `matchExpression` and `values` field. controller-runtime is looking at this markers and adding the field to manifets.

Reply:
"By default, Server-Side Apply treats custom resources as unstructured data. All keys are treated the same as struct fields, and all lists are considered atomic." Sounds conclusive to me (from your first reply)